### ISSUE TITLE
Cache WaniKani Vocab

### ISSUE THREAD

**Issue by domi2120** [ThePieMonster/WaniKanify-2.0#6]
Currently the extension loads the assignment + vocab data from WaniKani every time a website is opened / refreshed, this means that WaniKani's rate limit may swiftly become an issue (60 requests per minute, the extension easily does 14 or more requests at once), which also takes quite a bit of time, making the onLoad use of the extension rather impractical. There seems to already exist some form of caching mechanism, but i have not yet been able to see it actually being used (content.js line 2, 289, 549). Storing all data from Wanikani in localStorage takes about 9MB of storage, requiring the unlimited storage permission (perhaps some optimization could shrink that requirement. I haven't looked into that yet or whether it would suffice at higher levels, but i'd guess that if the mnemonics weren't stored it should be enough), but would avoid any issues with the rate limit and significantly speed up the extension (3-4 seconds). 

The numbers above assume lvl 30, so they may be higher or lower depending on the user.

TLDR:
WaniKani data is loaded on every website, this can swiftly cause issues with the rate limit.
This also causes the Extension to be rather slow due to large amounts of requests, making automatic execution impractical.
Caching all vocab data would require unlimited storage permissions, but this may be avoidable by omitting mnemonics.

Is there already a caching mechanism for this in place that i merely overlooked or if not, should there be one? (i'd be happy to implement it)

**Comment by ThePieMonster** [comment]
Thanks @domi2120. There is `function main(cache_local)` which is the primary driver at attempting to cache the content. As to whether it works as efficient as it should, probably not. 

My WK lvl is not that high currently so I don't have to many items to pull from for it be impactful enough on my tests.

Improving the caching is definitely a good idea, and if you had the time I'm sure it would be much appreciated. I would of course _first look at confirming and/or fixing the caching function_ before trying to downsize the total amount of data to try and remove the need for the unlimited storage permission. Either way there are multiple high level permissions that will be required so removing one of these will not speed up publishing time on Google end at all. Appreciate any contributions!

**Comment by ThePieMonster** [comment]
@domi2120 Just wanted to say thanks for taking the time to make these contributions. I'm sure the WaniKani users of this extension will appreciate the faster cache loading speeds. 👍
